### PR TITLE
 Update function-pointers, is-like, type-assertion BBEs 

### PR DESCRIPTION
**Pull request by ChinthakaSenanayaka**
## Automation tests
 - Unit tests 
   > Code coverage information N/A
 - Integration tests
   > Details about the test cases and coverage N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? N/A
 - Ran FindSecurityBugs plugin and verified report? N/A
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? N/A

### CODE REVIEW

File: examples/is-like/is_like.bal [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import ballerina/io;
```
Comment (pubudu91) [inline examples/is-like/is_like.bal]: We can't really have an example for the `is-like` check right? That's only an internal implementation detail. BBEs focus on explaining the usage of concrete language features. This instead should've been 2 examples on convert() and stamp() functions IMO.

File: examples/type-assertion/type_assertion.bal [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    int age = 0;
+};
+
+type Employee record {
+    string name = "";
+    int age = 0;
+    int empNo = 0;
+};
+
+function explainTypeAssertion(anydata emp) returns () {
```
Comment (MaryamZi) [inline examples/type-assertion/type_assertion.bal]: why do we need `returns ()`?

File: examples/type-assertion/type_assertion.bal [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    // There is no type casting happens here. Here, emp variable's type is asserted as `Employee` and assigns the same emp
+    // value to employee variable. This is safe type assertion.
```
Comment (MaryamZi) [inline examples/type-assertion/type_assertion.bal]: We don't need to say "There is no type casting happens here.", right?

We can just say 
"`emp` is asserted to be of type `Employee` and if successful the value is assigned to variable `employee`"

And, why are we calling this safe type assertion? IMO, the only safe type assertion for non basic types is if we do something like
```ballerina
Employee e = {};
Employee f = <Employee> e;
```

File: examples/type-assertion/type_assertion.bal [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    Employee employee = <Employee>emp;
+    io:println("Type asserted employee name: ", employee.name);
+
+    // This will panic since emp is inherently an Employee. `trap` is used to handle the error.
+    Person|error person = trap <Person>emp;
+    io:println("Type asserted person name or error: ", (person is Person) ? person.name : person.reason());
+}
+
+public function main() {
+    Employee emp2 = {name: "Speedy Gonzales", age: 4, empNo: 1};
```
Comment (MaryamZi) [inline examples/type-assertion/type_assertion.bal]: needs spaces after { and before }

File: examples/type-assertion/type_assertion.bal [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    explainTypeAssertion(emp2);
```
Comment (MaryamZi) [inline examples/type-assertion/type_assertion.bal]: shall we name the function something like `assertTypes` for consistency?

File: examples/type-assertion/type_assertion.description [hunk 6]
```diff
@@ -0,0 +1,4 @@
+// Type assertion is an exact type check. This will assert that given value is in asserted type. If it is, then assertion
+// assigns the provided value to the specified valriable, otherwise panic. In type assertion panic case, use `trap` to handle
+// the error. If the provided value that is being asserted E.g. type `any` or `anydata` then it is better to use `trap`
+// since it is not guaranteed to be safe.
```
Comment (MaryamZi) [inline examples/type-assertion/type_assertion.description]: How about something like
```suggestion
// Type assertion is a check for exact type. Type assertion checks if a given value's storage type is exactly a particular type.
// Type assertion results in panic, if assertion fails (i. e., the value is not of the asserted type).
```
IMO, we shouldn't describe panic, anydata, etc. here as concepts, instead we could use them in the example with short comments. WDYT?

File: examples/is-like/is_like.bal (hunk 1 above)
Comment (SupunS) [reply examples/is-like/is_like.bal]: +1 to have two samples. We can include them under 'types' section.

File: examples/is-like/is_like.bal [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    int age = 0;
+};
+
+type Employee record {
+    string name = "";
+    int age = 0;
+    int empNo = 0;
+};
+
+// Concept is-like contains `convert()` and `stamp()` functions. is-like is about converting runtime value's inherent storage type.
```
Comment (SupunS) [inline examples/is-like/is_like.bal]: I think we need to break the lines at col 80. Can you check?

File: examples/is-like/is_like.bal (hunk 1 above)
Comment (ChinthakaSenanayaka) [reply examples/is-like/is_like.bal]: Type-guard, type-convert examples are added. Mohan will add type-stamp

File: examples/type-assertion/type_assertion.bal (hunk 2 above)
Comment (ChinthakaSenanayaka) [reply examples/type-assertion/type_assertion.bal]: removed

File: examples/type-assertion/type_assertion.bal (hunk 4 above)
Comment (ChinthakaSenanayaka) [reply examples/type-assertion/type_assertion.bal]: done

File: examples/type-assertion/type_assertion.bal (hunk 5 above)
Comment (ChinthakaSenanayaka) [reply examples/type-assertion/type_assertion.bal]: done

File: examples/is-like/is_like.bal (hunk 7 above)
Comment (ChinthakaSenanayaka) [reply examples/is-like/is_like.bal]: File was removed

File: examples/type-convert/type_convert.bal [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+};
+
+type Employee record {
+    string name = "";
+    int age = 0;
+    int empNo = 0;
+};
+
+function explainTypeConvert(Employee emp) returns () {
+    // `convert()` creates a new value and changes its type without editing provided value's inherent type.
```
Comment (anupama-pathirage) [inline examples/type-convert/type_convert.bal]: Shall we start with The `convert() function...

File: examples/type-convert/type_convert.description [hunk 9]
```diff
@@ -0,0 +1 @@
+// `convert()` function creates a new runtime target value's storage with provided value and assign it to target variable.
```
Comment (anupama-pathirage) [inline examples/type-convert/type_convert.description]: Shall we start with The `convert() function...

File: examples/type-guard/type_guard.bal [hunk 10]
```diff
@@ -5,6 +5,38 @@ type KeyNotFoundError error<string, record { string key; }>;
...
+    string company;
+    string position;
+};
+
+function explainTypeTest(Person p) returns (Address?|Employment|error) {
+
+    // Person's employment is assigned to simple variable reference.
+    Employment? emp = p.employment;
+    if (p.address is Address) {
+        // type-test. thus this returns `Address?`
```
Comment (anupama-pathirage) [inline examples/type-guard/type_guard.bal]: Comment should be a valid statement. Check other places as well.

**Pull request merged**